internal/konnect/config_synchronizer.go

```diff
@@ -0,0 +1,165 @@
+package konnect
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/kong/go-database-reconciler/pkg/file"
+	"github.com/kong/go-kong/kong"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"github.com/kong/kubernetes-ingress-controller/v3/internal/adminapi"
+	"github.com/kong/kubernetes-ingress-controller/v3/internal/clients"
+	"github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane/deckerrors"
+	"github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane/sendconfig"
+	"github.com/kong/kubernetes-ingress-controller/v3/internal/metrics"
+)
+
+// DefaultConfigUploadPeriod is the default period between operations to upload Kong configuration to Konnect.
+var DefaultConfigUploadPeriod = 30 * time.Second
+
+// ConfigSynchronizer runs a loop to upload the traslated Kong configuration to Konnect in the given period.
+type ConfigSynchronizer struct {
+	logger                 logr.Logger
+	syncTicker             *time.Ticker
+	kongConfig             sendconfig.Config
+	clientsProvider        clients.AdminAPIClientsProvider
+	prometheusMetrics      *metrics.CtrlFuncMetrics
+	updateStrategyResolver sendconfig.UpdateStrategyResolver
+	configChangeDetector   sendconfig.ConfigurationChangeDetector
+	// REVIEW: use *file.Content to pass Kong configuration because it has Deepcopy method
+	// so we can easily copy config to prevent data race and long term locks.
```

Comment:
```suggestion
```
That sounds fine.

Some sort of read-only struct would probably be better but IDK that we want to breaking refactor the existing GDR struct to give it getter methods with the transforms we want (excluding consumers and whatever else). If we find we need to lock for longer than the update tick window to deepcopy that maybe becomes more of a need, but I don't know that anyone's hit that in practice.

Reply:
I don't think there is the possibility of too long time to take for `DeepCopy` because the operation happens in runtime and do not have network or IO dependency.